table-api/src/main/java/io/deephaven/api/Strings.java

```diff
@@ -27,52 +31,85 @@ public static String of(ColumnName columnName) {
+        if (invert) {
+            return "!" + encapsulate(inner);
         }
+        if (encapsulate) {
+            return encapsulate(inner);
+        }
+        return inner;
     }
 
-    public static String of(FilterNot not) {
-        return String.format("!(%s)", of(not.filter()));
+    public static String of(FilterComparison comparison) {
+        final String lhs = ofEncapsulated(comparison.lhs());
+        final String rhs = ofEncapsulated(comparison.rhs());
+        return String.format("%s %s %s", lhs, comparison.operator().javaOperator(), rhs);
+    }
+
+    public static String of(FilterComparison comparison, boolean encapsulate) {
+        final String inner = of(comparison);
+        return encapsulate ? encapsulate(inner) : inner;
     }
 
     public static String of(FilterIsNull isNull) {
-        return String.format("isNull(%s)", of(isNull.column()));
+        return String.format("isNull(%s)", of(isNull.expression()));
     }
 
-    public static String of(FilterIsNotNull isNotNull) {
-        return String.format("!isNull(%s)", of(isNotNull.column()));
+    public static String of(FilterIsNull isNull, boolean inverted) {
+        return (inverted ? "!" : "") + of(isNull);
     }
 
     public static String of(FilterOr filterOr) {
-        return filterOr.filters().stream().map(Strings::of)
-                .collect(Collectors.joining(") || (", "(", ")"));
+        return filterOr.filters().stream().map(Strings::ofEncapsulated).collect(Collectors.joining(" || "));
+    }
+
+    public static String of(FilterOr filterOr, boolean encapsulate) {
+        final String inner = of(filterOr);
+        return encapsulate ? encapsulate(inner) : inner;
     }
 
     public static String of(FilterAnd filterAnd) {
-        return filterAnd.filters().stream().map(Strings::of)
-                .collect(Collectors.joining(") && (", "(", ")"));
+        return filterAnd.filters().stream().map(Strings::ofEncapsulated).collect(Collectors.joining(" && "));
+    }
+
+    public static String of(FilterAnd filterAnd, boolean encapsulate) {
+        final String inner = of(filterAnd);
+        return encapsulate ? encapsulate(inner) : inner;
+    }
+
+    public static String of(FilterPattern pattern) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static String of(FilterPattern pattern, boolean encapsulate) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static String of(FilterQuick quick) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static String of(FilterQuick quick, boolean encapsulate, boolean inverted) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static String of(FilterMatches matches) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static String of(FilterMatches matches, boolean encapsulate, boolean inverted) {
+        // <ColumnName> [icase] [not] in <value 1>, <value 2>, ... , <value n>
```

Comment:
It _looks_ like this is reachable (and siblings too) by actually using filters as expressions? Isn't that in large part the point of the patch, that expressions can be expressed as filters?

Or will all such Filter types be guaranteed to be removed/rewritten by this point?

Reply:
The Strings class is (unfortunately) dual-purpose right now. It is used both for descriptive purposes, but also for transforming QSTs into strings for the purpose of falling back to filter & expression parsing.

In the case of `FilterMatches`, when used in the context of a `where`, it will flow through `io.deephaven.engine.table.impl.select.WhereFilterAdapter` and construct the appropriate `io.deephaven.engine.table.impl.select.MatchFilter`; in the case where `FilterMatches` is used as part of an expression, then that isn't currently supported by the engine and will hit this "UnsupportedOperationException" branch.

I think there are 3 potential ways forward:

* Leave as-is
* Print a pretty string here anyways, and (expect to) fail during formula parsing
* Separate out Strings for descriptions, vs Strings for engine parsing as separate utilities 

In the short-term, I'm hoping to avoid the work involved with the last bullet point. I'll create a ticket to follow-up in this regards.

Reply:
https://github.com/deephaven/deephaven-core/issues/3740